staging/src/k8s.io/kubectl/pkg/cmd/cp/filespec.go

```diff
@@ -0,0 +1,192 @@
+	Base() pathSpec
+	Clean() pathSpec
+	JoinElem(pathSpec) pathSpec
+	JoinFile(string) pathSpec
+	Glob() (matches []string, err error)
+	Prefix() pathSpec
+	StripShortcuts() pathSpec
+	// IsRelative returns true if dest is pointing outside the base directory, false otherwise.
+	IsRelative(string) bool
+}
+
+// localPath represents a client-native path, which will differ based
+// on the client OS, its methods will use path/filepath package which
+// is OS dependant
+type localPath struct {
+	file string
+}
+
+func newLocalPath(fileName string) pathSpec {
+	file := stripTrailingSlash(fileName)
+	return localPath{file: file}
+}
+
+func (p localPath) String() string {
+	return p.file
+}
+
+func (p localPath) Dir() pathSpec {
+	return newLocalPath(filepath.Dir(p.file))
+}
+
+func (p localPath) Base() pathSpec {
+	return newLocalPath(filepath.Base(p.file))
+}
+
+func (p localPath) Clean() pathSpec {
+	return newLocalPath(filepath.Clean(p.file))
+}
+
+func (p localPath) JoinElem(elem pathSpec) pathSpec {
+	return newLocalPath(filepath.Join(p.file, elem.String()))
+}
+
+func (p localPath) JoinFile(elem string) pathSpec {
+	return newLocalPath(filepath.Join(p.file, elem))
+}
+
+func (p localPath) Glob() (matches []string, err error) {
+	return filepath.Glob(p.file)
+}
+
+func (p localPath) Prefix() pathSpec {
+	return newLocalPath(stripLeadingSlash(p.file))
+}
+
+func (p localPath) StripShortcuts() pathSpec {
+	p = (p.Clean()).(localPath)
+	return newLocalPath(stripPathShortcuts(p.file))
+}
+
+func (p localPath) IsRelative(dest string) bool {
+	return false
+}
+
+// remotePath represents always UNIX path, its methods will use path
+// package which is always using `/`
+type remotePath struct {
+	file string
+}
+
+func newRemotePath(fileName string) pathSpec {
+	// we assume remote file is a linux container but we need to convert
+	// windows path separators to unix style for consistent processing
+	file := strings.ReplaceAll(stripTrailingSlash(fileName), `\`, "/")
+	return remotePath{file: file}
+}
+
+func (p remotePath) String() string {
+	return p.file
+}
+
+func (p remotePath) Dir() pathSpec {
+	return newRemotePath(path.Dir(p.file))
+}
+
+func (p remotePath) Base() pathSpec {
+	return newRemotePath(path.Base(p.file))
+}
+
+func (p remotePath) Clean() pathSpec {
+	return newRemotePath(path.Clean(p.file))
+}
+
+func (p remotePath) JoinElem(elem pathSpec) pathSpec {
+	return newRemotePath(path.Join(p.file, elem.String()))
+}
+
+func (p remotePath) JoinFile(elem string) pathSpec {
+	return newRemotePath(filepath.Join(p.file, elem))
+}
+
+func (p remotePath) Glob() (matches []string, err error) {
+	return nil, nil
+}
+
+func (p remotePath) Prefix() pathSpec {
+	return newLocalPath(stripLeadingSlash(p.file))
+}
+
+func (p remotePath) StripShortcuts() pathSpec {
+	p = (p.Clean()).(remotePath)
+	return newLocalPath(stripPathShortcuts(p.file))
+}
+
+func (p remotePath) IsRelative(dest string) bool {
+	relative, err := filepath.Rel(p.file, dest)
```

Comment:
Yes, fixed.

Reply:
Actually this is wrong, `filepath.Rel` should be client-side call, since that's invoked to ensure we can create a path relative file. So I left it as is, but moved to `localPath`, and also fixed all the places where during copying from pod I used `remotePath` instead of `localPath`.